variables.tf

```
@@ -37,7 +37,93 @@ variable "owners" {
 }
 
 variable "login_url" {
-  description = "(Optional) The URL where the service provider redirects the user to Azure AD to authenticate. Azure AD uses the URL to launch the application from Microsoft 365 or the Azure AD My Apps. When blank, Azure AD performs IdP-initiated sign-on for applications configured with SAML-based single sign-on."
+  description = "The URL where the service provider redirects the user to Azure AD to authenticate. Azure AD uses the URL to launch the application from Microsoft 365 or the Azure AD My Apps. When blank, Azure AD performs IdP-initiated sign-on for applications configured with SAML-based single sign-on."
   type        = string
   default     = null
 }
+
+variable "web_homepage_url" {
+  description = "Application home page."
+  type        = string
+  default     = null
+}
+
+variable "web_logout_url" {
+  description = "The URL that will be used to sign out."
+  type        = string
+  default     = null
+}
+
+variable "web_redirect_uris" {
+  description = "A set of URLs where OAuth 2.0 autorization codes and access tokens are sent."
+  type        = set(string)
+
+  validation {
+    condition     = alltrue([for uri in var.web_redirect_uris : can(regex("^https://|^ms-appx-web://|^http://localhost", uri))])
+    error_message = "All URIs must be valid HTTPS URLs excluding localhost."
+  }
+
+  default = []
+}
+
+variable "public_client_redirect_uris" {
+  description = "A set of URLs where OAuth 2.0 autorization codes and access tokens are sent."
+  type        = set(string)
+
+  validation {
+    condition     = alltrue([for uri in var.public_client_redirect_uris : can(regex("^https://|^ms-appx-web://", uri))])
+    error_message = "All URIs must be valid HTTPS or ms-appx-web URLs."
+  }
+
+  default = []
+}
+
+variable "access_token_issuance_enabled" {
+  description = "Should the application be allowed to request an access token?"
+  type        = bool
+  default     = false
+}
+
+variable "id_token_issuance_enabled" {
+  description = "Should the application be allowed to request an ID token?"
+  type        = bool
+  default     = false
+}
+
+variable "required_resource_access" {
+  description = "List of required resource access blocks"
+  type = list(object({
+    resource_app_id = string
+    resource_access = list(object({
+      id   = string
+      type = string
+    }))
+  }))
+
+  validation {
+    condition     = alltrue([for resource in var.required_resource_access : alltrue([for access in resource.resource_access : access.type == "Role" || access.type == "Scope"])])
+    error_message = "Type must be either \"Role\" or \"Scope\""
+  }
+
+  default = []
+}
+
+variable "oauth2_permission_scopes" {
+  description = "List of required oauth permission scope"
+  type = list(object({
+    admin_consent_description  = string
+    admin_consent_display_name = string
+    enabled                    = bool
+    type                       = optional(string)
```

This should either have a default value or not be marked as optional, or else the custom validation rule will strike by default 🙂 

```suggestion
    type                       = optional(string)
```